### PR TITLE
Union type inference

### PR DESCRIPTION
**Pull request by benlesh**
This is work done to solve for #4387

- Adds a `ObservedValueOf` type that is used to infer the type *out of an* `ObservableInput<any>`. This allows us to infer types properly from situations where projection functions in flattening operators return union types. This is a common thing in redux-observable and/or ngrx effects.

For example: this would have failed before this change...

```ts
const source = of(1, 2, 3).pipe(concatMap(() => Math.random() > 0.5 ? of(123) : of('test'));
```

... because TS wouldn't have been able to infer the type properly from the projection function. With this PR it can.

### CODE REVIEW

**Comment by benlesh** [comment]
Looks like a few errors in CI. I'll get back to this PR soon.

File: spec-dtslint/operators/switchMap-spec.ts [hunk 1]
```diff
@@ -18,11 +18,11 @@ it('should infer correctly by using the resultSelector second parameter', () =>
 });
 
 it('should support a resultSelector that takes an inner index', () => {
-  const o = of(1, 2, 3).pipe(switchMap(p => of(Boolean(p)), (a, b, innnerIndex) => a)); // $ExpectType Observable<number>
+  const o = of(1, 2, 3).pipe(switchMap(p => of(Boolean(p)), a => a)); // $ExpectType Observable<number>
 });
 
 it('should support a resultSelector that takes an inner and outer index', () => {
-  const o = of(1, 2, 3).pipe(switchMap(p => of(Boolean(p)), (a, b, innnerIndex, outerIndex) => a)); // $ExpectType Observable<number>
+  const o = of(1, 2, 3).pipe(switchMap(p => of(Boolean(p)), a => a)); // $ExpectType Observable<number>
```
[the remark this answers is not in the document]
Comment (benlesh) [reply spec-dtslint/operators/switchMap-spec.ts]: Yep. Late night mistake.

File: src/internal/operators/catchError.ts [hunk 2]
```diff
@@ -5,10 +5,11 @@ import {Observable} from '../Observable';
 import {OuterSubscriber} from '../OuterSubscriber';
 import { InnerSubscriber } from '../InnerSubscriber';
 import {subscribeToResult} from '../util/subscribeToResult';
-import {ObservableInput, OperatorFunction, MonoTypeOperatorFunction} from '../types';
+import {ObservableInput, OperatorFunction, MonoTypeOperatorFunction, ObservedValueOf} from '../types';
 
-export function catchError<T>(selector: (err: any, caught: Observable<T>) => never): MonoTypeOperatorFunction<T>;
```
[the remark this answers is not in the document]
Comment (cartant) [reply src/internal/operators/catchError.ts]: Yeah, I can see, now, that the explicit `never` in the removed is effected from within the `ObservedValueOf` conditional type.

File: spec-dtslint/observables/combineLatest-spec.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+const a = of(new A());
+const b = of(new B());
+const c = of(new C());
+const d = of(new D());
+const e = of(new E());
+const f = of(new F());
+const g = of(new G());
+
+it('should accept 1 param', () => {
+  // Since this is an uncommon case, we're just allowing A[]
```
[the remark this answers is not in the document]
Comment (benlesh) [reply spec-dtslint/observables/combineLatest-spec.ts]: Outdated comment. :) Thanks for catching that.

File: src/internal/operators/mergeAll.ts [hunk 4]
```diff
@@ -57,6 +55,6 @@ export function mergeAll<T>(concurrent?: number): OperatorFunction<ObservableInp
  * @method mergeAll
  * @owner Observable
  */
-export function mergeAll<T>(concurrent: number = Number.POSITIVE_INFINITY): MonoTypeOperatorFunction<T> {
-  return mergeMap<T, T>(identity as (value: T, index: number) => ObservableInput<T>, concurrent);
+export function mergeAll<T>(concurrent: number = Number.POSITIVE_INFINITY): OperatorFunction<ObservableInput<T>, T> {
```
[the remark this answers is not in the document]
Comment (benlesh) [reply src/internal/operators/mergeAll.ts]: The old type was wrong here... you can't perform `mergeAll` on anything but an `Observable<ObservableInput<T>>`.

**Pull request merged**